server/schedulers/balance_region.go

```diff
@@ -241,3 +261,130 @@ func (s *balanceRegionScheduler) transferPeer(cluster opt.Cluster, region *core.
+	defaultTTL time.Duration
+	threshold  int
+}
+
+func newHitsStoreBuilder(defaultTTL time.Duration, threshold int) *hitsStoreBuilder {
+	return &hitsStoreBuilder{
+		hits:       make(map[string]*record),
+		ttl:        make(map[uint64]time.Duration),
+		defaultTTL: defaultTTL,
+		threshold:  threshold,
+	}
+}
+
+func (h *hitsStoreBuilder) getKey(source, target *core.StoreInfo) string {
+	if source == nil {
+		return ""
+	}
+	key := fmt.Sprintf("s%d", source.GetID())
+	if target != nil {
+		key = fmt.Sprintf("%s->t%d", key, target.GetID())
+	}
+	return key
+}
+
+func (h *hitsStoreBuilder) filter(source, target *core.StoreInfo) bool {
+	key := h.getKey(source, target)
+	ttl := h.getTTL(source, target)
+	if key == "" {
+		return false
+	}
+	if item, ok := h.hits[key]; ok {
+		if time.Since(item.lastTime) > ttl {
+			delete(h.hits, key)
+		}
+		if time.Since(item.lastTime) <= ttl && item.count >= h.threshold {
+			log.Debug("skip the the store", zap.String("scheduler", BalanceRegionName), zap.String("filter-key", key))
+			return true
+		}
+	}
+	return false
+}
+
+func (h *hitsStoreBuilder) remove(source, target *core.StoreInfo) {
+	key := h.getKey(source, target)
+	if _, ok := h.hits[key]; ok && key != "" {
+		delete(h.hits, key)
+	}
+}
+
+func (h *hitsStoreBuilder) put(source, target *core.StoreInfo) {
+	key := h.getKey(source, target)
+	ttl := h.getTTL(source, target)
+	if key == "" {
+		return
+	}
+	if item, ok := h.hits[key]; ok {
+		if time.Since(item.lastTime) >= ttl {
+			item.count = 0
+		} else {
+			item.count++
+		}
+		item.lastTime = time.Now()
+	} else {
+		item := &record{lastTime: time.Now()}
+		h.hits[key] = item
+	}
+}
+
+func (h *hitsStoreBuilder) buildSourceFilter(scope string, cluster opt.Cluster) filter.Filter {
+	f := filter.NewBlacklistStoreFilter(scope, filter.BlacklistSource)
+	h.updateTTL(cluster.GetAllStoresLimit())
+	for _, source := range cluster.GetStores() {
+		if h.filter(source, nil) {
+			f.Add(source.GetID())
+		}
+	}
+	return f
+}
+
+func (h *hitsStoreBuilder) buildTargetFilter(scope string, cluster opt.Cluster, source *core.StoreInfo) filter.Filter {
+	f := filter.NewBlacklistStoreFilter(scope, filter.BlacklistTarget)
+	h.updateTTL(cluster.GetAllStoresLimit())
+	for _, target := range cluster.GetStores() {
+		if h.filter(source, target) {
+			f.Add(target.GetID())
+		}
+	}
+	return f
+}
+
+func (h *hitsStoreBuilder) getTTL(source, target *core.StoreInfo) time.Duration {
+	sourceTTL := h.readTTL(source)
+	targetTTL := h.readTTL(target)
+	if sourceTTL < targetTTL {
+		return sourceTTL
+	}
+	return targetTTL
+}
+
+func (h *hitsStoreBuilder) readTTL(store *core.StoreInfo) time.Duration {
+	ttl := h.defaultTTL
+	if store != nil {
+		if sourceTTL, ok := h.ttl[store.GetID()]; ok && sourceTTL < ttl {
+			ttl = sourceTTL
+		}
+	}
+	return ttl
+}
+
+func (h *hitsStoreBuilder) updateTTL(limits map[uint64]float64) {
+	for storeID, limit := range limits {
+		h.ttl[storeID] = time.Second * time.Duration(limit*hitLimitFactor)
```

Comment:
How about add some comments to explain why `time.Second * time.Duration(limit*hitLimitFactor)`?